### DISCUSSION TITLE
How to locate SDK in vscode ?

### DISCUSSION THREAD

**Discussion by chapuyj** [jdx/mise#3416] [Q&A]
Hi!

I migrate to mise and start to love it. But I have a problem and I can't find a solution. The Flutter extension in vscode asks me to locate the path to the Flutter SDK ("dart.flutterSdkPath" in settings.json). But I don't know what to put, I can't get it to work.

I tried to set shims path (`/Users/chapuyj/.local/share/mise/shims/`), but it doesn't work:
```
mise ERROR No version is set for shim: flutter
Set a global default version with one of the following:
mise use -g flutter@3.24.5-stable
mise ERROR Run with --verbose or MISE_VERBOSE=1 for more information
exit code 1
```

In iTerm, I can use `flutter` from my project folder (which has a `mise.toml`). From the vscode terminal, I can also use the `flutter` command.

In `~/.zshrc ` I activate `mise`:
```
if [[ "$TERM_PROGRAM" == "vscode" ]]; then
  eval "$($HOME/.local/bin/mise activate zsh --shims)"
else
  eval "$($HOME/.local/bin/mise activate zsh)"
fi
```

I don't know what to do, maybe there's something I don't understand.

I'm looking for help to understand better, and to get the extension to work. Thanks :)

**Comment by jdx** [comment]
that error basically means the shim is being called when the process isn't in that directory. You could set a global version like the error says or figure out how to chdir into your project directory somehow. I haven't used vscode in a long time so no idea how to do that.

**Reply by chapuyj** [reply]
Yes, it's a possibility I've been thinking about. But ideally I'm looking for a solution that allows me to stay “dynamic” according to the current project.

**Comment by syhol** [comment]
Try setting `dart.flutterSdkPath` to `mise where flutter` to get the sdk path. It's not looking for the binary path.

**Reply by chapuyj** [reply]
That's what I'm doing right now, to make it work. I set `/Users/chapuyj/.local/share/mise/installs/flutter/3.24.5-stable/bin/` in `.vscode/settings.json`. So it's only set for this project. But not perfect : if I need to change the sdk version, I have to change it with mise and then in settings.json.

Unless there's a way to put the command execution in the settings (rather than the hard-coded value)? It doesn't seem possible at first look.

**Comment by hverlin** [comment]
You can try with https://github.com/hverlin/mise-vscode, it will automatically configure `flutter` based on the configuration. This will do the same as @syhol suggested but automatically. 

`dart` and `flutter` extension don't seem to work with shims, unfortunately. (https://github.com/hverlin/mise-vscode/wiki/Supported-extensions)

**Reply by chapuyj** [reply]
Thanks. If I understand, the page said Flutter/Dart `Does not work with shims or symlinks`. So, I must edit my `~/.zshrc` to not use shims finally (as opposed to the mise documentation for IDEs). Then install mise-vscode, configure the mise-vscode to not use shims, and it's do magic?

 By the way, I'd like to understand what's going on behind the scenes. Will mise-vscode automatically generate the path in `.vscode/settings.json` each time the project is opened?)

**Reply by hverlin** [reply]
You do not need to change anything in your zshrc. The extension works even if you did not activate `mise` or don't have shims enabled. It's using `mise ls --json` to know how tools are configured for your project.

I will clarify the documentation, but this means that the extension cannot use a shim to configure the extension. (as you have noticed, it needs the install dir and not the bin path). 

The symlink option is only there for people who like to share their `.vscode/settings.json` as it symlinks the path to the tool relative to your project (this way the settings.json file works for everyone in your team if everyone is also using `mise`.

> Will mise-vscode automatically generate the path in .vscode/settings.json each time the project is opened?

yeah, actually anytime you update your `mise` toml file, it will update the settings.json file to configure all extensions it knows about. You can opt out of this, of course.

**Reply by hverlin** [reply]
Here is what it does (note that I already installed the latest version of flutter with mise before recording as it takes ~1 min otherwise with the asdf plugin)

https://github.com/user-attachments/assets/e6cdaf60-b7fb-40d2-b2ad-46a9d6d837ae

**Reply by chapuyj** [reply]
Thanks for details and video! I'll try it later this day.

By the way. On the video, I see that it changes automatically because the modification is made from VSCode (and therefore the extension is active).
Does the extension also check and update when the project is opened on VSCode? (if, for example, I make a change outside vscode, or if another person changes the version and I git pull).

**Reply by hverlin** [reply]
There is a file watcher for mise configuration files. (e.g. if you run `mise use` and VSCode is open, it will update the settings)

However, in case it does not trigger, you can always trigger this action manually using the command palette. 
![image](https://github.com/user-attachments/assets/eafe45d2-cac4-4914-afa4-e274c83f83af)

(there are still edge cases where it fails - e.g. if you open a project and have an invalid mise configuration file, it will not register the file watcher, but this will get fixed in the next releases)
